README.md
# Snowplow AWS Lambda source

[ ![Build Status] [travis-image] ] [travis]
[ ![Release] [release-image] ] [releases]
[ ![License] [license-image] ] [license]

Snowplow AWS Lambda source is a tool to track [AWS Lamba event sources](http://docs.aws.amazon.com/lambda/latest/dg/intro-core-components.html#intro-core-components-event-sources) in [Snowplow](http://snowplowanalytics.com/).

## Currently supported sources

* S3 Bucket operations
   * Depending on how you configure the lambda, this can include when a file is created, and when a file is deleted in a single or multiple S3 buckets
   
## Deployment

For a detailed set-up guide, see [here] [setup]. 
   
## Developer quickstart

Snowplow AWS Lambda source is written in Java 8 and uses gradle as a build tool. To get started quickly - a Vagrant set-up is also provided.

### Using Vagrant

The below commands will set up a Vagrant box provisioning Java 8 and gradle.

```{bash}
git clone git@github.com:snowplow/snowplow-aws-lambda-source.git
cd ./snowplow-aws-lambda-source
vagrant up && vagrant ssh
cd /vagrant
./gradlew test
```

### Using the gradle wrapper

```{bash}
git clone git@github.com:snowplow/snowplow-aws-lambda-source.git
cd ./snowplow-aws-lambda-source
./gradlew test
```

## Find out more

| Technical Docs                  | Setup Guide               | Roadmap                 | Contributing                      |
|---------------------------------|---------------------------|-------------------------|-----------------------------------|
| ![i1] [techdocs-image]          | ![i2] [setup-image]       | ![i3] [roadmap-image]   | ![i4] [contributing-image]        |
| **[Technical Docs] [techdocs]** | **[Setup Guide] [setup]** | **[Roadmap] [roadmap]** | **[Contributing] [contributing]** |

### Copyright and license

Snowplow is copyright 2016 Snowplow Analytics Ltd.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this software except in compliance with the License.

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

[travis-image]: https://travis-ci.org/snowplow/snowplow-aws-lambda-source.png?branch=master
[travis]: http://travis-ci.org/snowplow/snowplow-aws-lambda-source

[license-image]: http://img.shields.io/badge/license-Apache--2-blue.svg?style=flat
[license]: http://www.apache.org/licenses/LICENSE-2.0

[release-image]: https://img.shields.io/badge/release-0.1.0-orange.svg?style=flat
[releases]: https://github.com/snowplow/snowplow/releases

[techdocs]:     https://github.com/snowplow/snowplow/wiki/AWS-Lambda-source
[setup]:        https://github.com/snowplow/snowplow/wiki/AWS-Lambda-source-setup
[roadmap]:      https://github.com/snowplow/snowplow/wiki/AWS-Lambda-source-roadmap
[contributing]: https://github.com/snowplow/snowplow/wiki/AWS-Lambda-source-contributing

[techdocs-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/techdocs.png
[setup-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/setup.png
[roadmap-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/roadmap.png
[contributing-image]: https://d3i6fms1cm1j0i.cloudfront.net/github/images/contributing.png


CHANGELOG
Version 0.1.0 (2016-05-13)
--------------------------
Initial release


deploy/test_deploy.py
import deploy
import unittest
import json
import re
import subprocess
import os.path
import json

class MockExecutor(deploy.Executor):
   
    def __init__(self, return_code=0, so=""):
        self._stdout = so
        self._return_code = return_code
        self._exec_calls = []
        
    def exec_output(self, args):
        self._exec_calls.append(args)
        if (self._return_code != 0):
            raise subprocess.CalledProcessError('potato', 'banana', 'swift')
        return self._stdout

    def call(self, args):
        self._exec_calls.append(args)
        if (self._return_code != 0):
            raise subprocess.CalledProcessError('man', 'bear', 'pig')
        return self._return_code

    def get_call(self, index):
        return self._exec_calls[index]
        
    def get_call_count(self):
        return len(self._exec_calls)
        
class TestConfiguration(unittest.TestCase):
        
    def test_file_does_not_exist(self):
        with self.assertRaises(IOError) as m: deploy.Configuration("asdfasfdaf,afsdfa").conf
        
    def test_empty_str(self):
        with self.assertRaises(ValueError) as m: deploy.Configuration.parse_str('')
        
    def test_missing_snowplow(self):
        co = """
s3:
    buckets:
        - abc
        - 123
"""
        with self.assertRaises(ValueError) as m: deploy.Configuration.parse_str(co)
        
    def test_missing_snowplow_collector(self):
        co = """
snowplow:
    #collector: http://helloworld.com
    app_id: com.lambda
s3:
    buckets:
        - abc
        - 123
"""
        with self.assertRaises(ValueError) as m: deploy.Configuration.parse_str(co)
        
      
    def test_empty_snowplow_collector(self):
        co = """
snowplow:
    collector:
    app_id: com.lambda
s3:
    buckets:
        - abc
        - 123
"""
        with self.assertRaises(ValueError) as m: deploy.Configuration.parse_str(co)
        
    def test_missing_s3(self):
        co = """
snowplow:
    collector: http://helloworld.com
    app_id: com.lambda
"""
        with self.assertRaises(ValueError) as m: deploy.Configuration.parse_str(co)
        
    def test_missing_s3_buckets(self):
        co = """
snowplow:
    collector: http://helloworld.com
    app_id: com.lambda
s3:
    hello: world
"""
        with self.assertRaises(ValueError) as m: deploy.Configuration.parse_str(co)
        
    def test_no_s3_buckets(self):
        co = """
snowplow:
    collector: http://helloworld.com
    app_id: com.lambda
s3:
    buckets:
"""
        with self.assertRaises(ValueError) as m: deploy.Configuration.parse_str(co)
        
    def test_empty_s3_buckets(self):
        co = """
snowplow:
    collector: http://helloworld.com
    app_id: com.lambda
s3:
    buckets:
        -
        -
"""
        with self.assertRaises(ValueError) as m: print str(deploy.Configuration.parse_str(co))
        
        
    def test_no_appid(self):
        co = """
snowplow:
    collector: http://helloworld.com
s3:
    buckets:
        - hello
        - world
"""
        with self.assertRaises(ValueError) as m: print str(deploy.Configuration.parse_str(co))
        
    def test_blank_appid(self):
        co = """
snowplow:
    collector: http://helloworld.com
    app_id: 
s3:
    buckets:
        - hello
        - world
"""
        with self.assertRaises(ValueError) as m: print str(deploy.Configuration.parse_str(co))
    
class TestShell(unittest.TestCase):
        
    def test_is_aws_cli(self):
        self.assertTrue(deploy.Shell.contains_aws_cli(MockExecutor(0)))
        self.assertFalse(deploy.Shell.contains_aws_cli(MockExecutor(1)))
        mock_executor = MockExecutor(0)
        deploy.Shell.contains_aws_cli(mock_executor)
        self.assertEquals(mock_executor.get_call(0), ['which', 'aws-cli'])

class TestS3Operations(unittest.TestCase):

   def test_get_bucket_list(self):
        executor = MockExecutor(0, '["hello", "world"]')
        buckets_json = deploy.S3.get_bucket_list(executor)
        self.assertEquals(executor.get_call(0), ['aws', 's3api', 'list-buckets', '--query', 'Buckets[].Name'])
                    
   def test_get_buckets(self):
        sample_bucket_list = """
          [
            "zero2hadoop", 
            "hadoopy-ed", 
            "leaky" 
          ]
        """
        mock_executor = MockExecutor(0, sample_bucket_list)
        b = deploy.S3.get_bucket_list(mock_executor)
        self.assertTrue(len(b)>0)
        buckets = deploy.S3.get_buckets_from_list(b)
        self.assertEquals(len(buckets), 3)  
    
   def test_available_buckets(self):
        sample_bucket_list = """
          [
            "zero2hadoop", 
            "hadoopy-ed", 
            "leaky" 
          ]
        """
        mock_executor = MockExecutor(0, sample_bucket_list)
        op = deploy.S3(mock_executor)
        self.assertEquals(op.available_buckets, ['zero2hadoop', 'hadoopy-ed', 'leaky'])               
                
   def test_attach_buckets(self):
        with self.assertRaises(ValueError) as cm: deploy.S3.add_bucket_notifications('abc', '1323', list(), 'hello')         
        
   def test_add_bucket_notifications(self):
        executor = MockExecutor(0)
        buckets = ['apple','orange']
        r = deploy.S3.add_bucket_notifications('lambda-func', 'arn123', buckets, '1234', executor)
        self.assertEquals(r, None)
        self.assertEquals(executor.get_call_count(), len(buckets)*2)
        expected_notification_conf = '{"CloudFunctionConfiguration":{"CloudFunction":"arn123","Events":["s3:ObjectCreated:*","s3:ObjectRemoved:*"]}}'
        self.assertFalse(None == json.loads(expected_notification_conf))
        
        expected_first_call = ['aws', 'lambda', 'add-permission', '--function-name', 'lambda-func', '--action', 'lambda:InvokeFunction', 
                               '--principal', 's3.amazonaws.com', '--source-arn', 'arn:aws:s3:::apple', '--source-account', '1234', '--statement-id', 'Id-1']
        expected_second_call = ['aws', 's3api', 'put-bucket-notification', '--bucket', 'apple', '--notification', expected_notification_conf]
        
        expected_third_call = ['aws', 'lambda', 'add-permission', '--function-name', 'lambda-func', '--action', 'lambda:InvokeFunction', 
                               '--principal', 's3.amazonaws.com', '--source-arn', 'arn:aws:s3:::orange', '--source-account', '1234', '--statement-id', 'Id-2']
        expected_fourth_call = ['aws', 's3api', 'put-bucket-notification', '--bucket', 'orange', '--notification', expected_notification_conf]
        
        self.assertEquals(executor.get_call(0), expected_first_call)
        self.assertEquals(executor.get_call(1), expected_second_call)
        self.assertEquals(executor.get_call(2), expected_third_call)
        self.assertEquals(executor.get_call(3), expected_fourth_call)
        
class TestAWSUtils(unittest.TestCase):
              
    def test_get_aws_account_id(self):
        executor = MockExecutor(0, "12345\n")
        id = deploy.AWS.get_account_id(executor)
        self.assertEquals(executor.get_call(0), ['aws', 'ec2', 'describe-security-groups', '--group-names', 'Default', '--query', 'SecurityGroups[0].OwnerId', '--output', 'text']) 
        self.assertEquals('12345', id)
        with self.assertRaises(subprocess.CalledProcessError) as m: deploy.AWS.get_account_id(MockExecutor(1))                              

class TestIAM(unittest.TestCase):

    def test_get_arn_from_policy(self):
        pol = """
                {
                    "Policy": {
                        "PolicyName": "***", 
                        "CreateDate": "2016-05-19T09:14:56.204Z", 
                        "AttachmentCount": 0, 
                        "IsAttachable": true, 
                        "PolicyId": "***", 
                        "DefaultVersionId": "v1", 
                        "Path": "/", 
                        "Arn": "arn:aws:iam::123:policy/snowplow-s3-source", 
                        "UpdateDate": "2016-05-19T09:14:56.204Z"
                    }
                } 
              """
        arn = deploy.IAM.get_arn_from_policy(pol)
        self.assertEquals(arn, 'arn:aws:iam::123:policy/snowplow-s3-source')
        
    def test_get_arn_from_role(self):
        pol = """
           {
                "Role": {
                    "AssumeRolePolicyDocument": {
                        "Version": "2012-10-17", 
                        "Statement": [
                            {
                                "Action": "sts:AssumeRole", 
                                "Principal": {
                                    "Service": "lambda.amazonaws.com"
                                }, 
                                "Effect": "Allow", 
                                "Sid": ""
                            }
                        ]
                    }, 
                    "RoleId": "***", 
                    "CreateDate": "2016-05-19T09:26:07.305Z", 
                    "RoleName": "snowplow-s3-source", 
                    "Path": "/", 
                    "Arn": "arn:aws:iam::123:role/snowplow-s3-source"
                }
            }        
        """
        arn = deploy.IAM.get_arn_from_role(pol)
        self.assertEquals(arn, 'arn:aws:iam::123:role/snowplow-s3-source')
               
    def test_create_lambda_policy(self):
        sample_json = '{"hello":"world"}'
        executor = MockExecutor(0, sample_json)
        r = deploy.IAM.create_lambda_policy('lambda.json',executor)
        self.assertEquals(executor.get_call(0), ['aws', 'iam', 'create-policy', '--policy-name', 'snowplow-s3-source', '--policy-document', 'file://lambda.json'])
        self.assertEquals(r, sample_json)
        
    def test_create_lambda_role(self):
        sample_json = '{"hello":"world"}'
        executor = MockExecutor(0, sample_json)
        r = deploy.IAM.create_lambda_role('sp-s3','tp.json', executor)
        self.assertEquals(executor.get_call(0), ['aws', 'iam', 'create-role', '--role-name', 'sp-s3', '--assume-role-policy-document', 'file://tp.json'])
        self.assertEquals(r, sample_json)
    
    def test_attach_policy_to_role(self):
        sample_json = '{"hello":"world"}'
        executor = MockExecutor(0, sample_json)
        r = deploy.IAM.attach_policy_to_role('hello', 'world', executor)
        self.assertEquals(executor.get_call(0), ['aws', 'iam', 'attach-role-policy','--role-name=hello', '--policy-arn', 'world'])
        self.assertEquals(r, None)

class TestLambda(unittest.TestCase):    
           
    def test_get_lambda_arn_from_lambda(self):
        res = """
            {
                "CodeSha256": "G4L8Rcdlc0vLY69GUOwzujApBge9ViRkHoxtMuii9Tg=", 
                "FunctionName": "snowplow-aws-lambda", 
                "CodeSize": 38563260, 
                "MemorySize": 128, 
                "FunctionArn": "arn:aws:lambda:eu-west-1:123:function:snowplow-aws-lambda", 
                "Version": "$LATEST", 
                "Role": "arn:aws:iam::123:role/snowplow-s3-source", 
                "Timeout": 120, 
                "LastModified": "2016-05-19T09:55:08.102+0000", 
                "Handler": "com.snowplowanalytics.sources.awslambda.Main::handleRequest", 
                "Runtime": "java8", 
                "Description": " "
            }
            """
        arn = deploy.Lambda.get_lambda_function_arn(res)
        self.assertEquals(arn, 'arn:aws:lambda:eu-west-1:123:function:snowplow-aws-lambda')
        
    def test_create_lambda_function(self):
        expected = ['aws', 'lambda', 'create-function', '--function-name', 'lambda-name', '--runtime',
                    'java8', '--role', 'role-arn-123', '--handler', 'com.snowplowanalytics.sources.awslambda.Main::handleRequest',
                    '--zip-file', 'fileb://lambda.zip', '--description', ' http://abc.com', '--timeout', '120', '--memory-size', '512']
        executor = MockExecutor(0, "stdout")
        r = deploy.Lambda.create_lambda_function('lambda-name', 'lambda.zip', 'role-arn-123', 'http://abc.com', executor)
        self.assertEquals(executor.get_call(0), expected)
        self.assertEquals(r, "stdout")      
        
        
class MiscTest(unittest.TestCase):

    def test_get_lambda_zip_dir(self):
        try:
            os.utime("test.zip", None)
        except:
            open("test.zip", 'a').close()
            
        self.assertEquals("test.zip", deploy.get_lambda_zip_filename("abc"))
        self.assertEquals("test.zip", deploy.get_lambda_zip_filename(["abc", "test.zip"]))

        os.remove("test.zip")
        
        with self.assertRaises(ValueError) as m: deploy.get_lambda_zip_filename(["abc", "deploy.py"])
        with self.assertRaises(ValueError) as m: deploy.get_lambda_zip_filename(["abc", "abc.zip"]) 
        with self.assertRaises(ValueError) as m: deploy.get_lambda_zip_filename(["abc"])        
        
    def test_make_lambda_config_desc(self):
        r = deploy.make_lambda_config_desc("collector_url", "app_id")
        self.assertEquals("""{"collector_url":"collector_url","app_id":"app_id"}""", r)
        json.loads(r)
        with self.assertRaises(ValueError) as m: deploy.make_lambda_config_desc(None, "abc")
        with self.assertRaises(ValueError) as m: deploy.make_lambda_config_desc("abc", None) 
        with self.assertRaises(ValueError) as m: deploy.make_lambda_config_desc(None, None)
                
if __name__ == '__main__':
    unittest.main()
    

src/test/java/com/snowplowanalytics/sources/awslambda/MainTest.java
/*
 * Copyright (c) 2016 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0, and
 * you may not use this file except in compliance with the Apache License
 * Version 2.0.  You may obtain a copy of the Apache License Version 2.0 at
 * http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Apache License Version 2.0 is distributed on an "AS
 * IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied.  See the Apache License Version 2.0 for the specific language
 * governing permissions and limitations there under.
 */
package com.snowplowanalytics.sources.awslambda;

import com.snowplowanalytics.snowplow.tracker.http.HttpClientAdapter;
import com.snowplowanalytics.snowplow.tracker.payload.SelfDescribingJson;
import com.squareup.okhttp.OkHttpClient;
import org.junit.Assert;
import org.junit.Test;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.IntStream;
import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.*;

public class MainTest {

    @Test
    public void testGetRegion() throws Exception {
        String mockArn = "arn:aws:elasticbeanstalk:us-east-1:123456789012:environment/My App/MyEnvironment";
        Assert.assertEquals("us-east-1", Main.getRegion(mockArn));
    }

    @Test (expected = IllegalArgumentException.class)
    public void testGetRegionNull() {
        Main.getRegion(null);
    }

    @Test (expected = IllegalArgumentException.class)
    public void testGetRegionEmpty() {
        Main.getRegion("");
    }

    @Test (expected = IllegalArgumentException.class)
    public void testGetRegionShort() {
        Main.getRegion("1:2:3:"); // not enough ARN
    }

    @Test
    public void testGetAdapter() {
        HttpClientAdapter adapter = Main.getClient("http://snowplowanalytics.com");
        assertEquals("http://snowplowanalytics.com", adapter.getUrl());
        OkHttpClient client = (OkHttpClient)adapter.getHttpClient();
        assertEquals(5000, client.getReadTimeout());
        assertEquals(5000, client.getConnectTimeout());
        assertEquals(5000, client.getWriteTimeout());
    }

    @Test
    public void testTrackEvents() {
        HttpClientAdapter adapter = mock(HttpClientAdapter.class);
        when(adapter.post(any())).thenReturn(200);

        Map<String,String> testEventContent = new HashMap<>();
        testEventContent.put("hello", "world");

        List<SelfDescribingJson> testEvents = new ArrayList<>();
        IntStream.rangeClosed(1,100).forEach(i -> {
            testEvents.add(new SelfDescribingJson("iglu:com.acme/event/jsonschema/1-0-0", testEventContent));
        });

        Main m = new Main();
        m.trackEvents(adapter, testEvents, "main", "com.snowplowanalytics.test");

        verify(adapter).post(any());
    }

    @Test (expected = RuntimeException.class)
    public void testFailuresToTrack() {
            HttpClientAdapter adapter = mock(HttpClientAdapter.class);
            when(adapter.post(any())).thenReturn(503);

            Map<String,String> testEventContent = new HashMap<>();
            testEventContent.put("hello", "world");

            List<SelfDescribingJson> testEvents = new ArrayList<>();
            IntStream.rangeClosed(1,100).forEach(i -> {
                testEvents.add(new SelfDescribingJson("iglu:com.acme/event/jsonschema/1-0-0", testEventContent));
            });

            Main m = new Main();
            m.trackEvents(adapter, testEvents, "main", "com.snowplowanalytics.test");

            verify(adapter).post(any());
    }

    @Test
    public void testToEvent() {
        SelfDescribingJson s = new SelfDescribingJson("iglu:com.acme/event/jsonschema/1-0-0", new Object());
        assertEquals(s.toString(), Main.toEvent("iglu:com.acme/event/jsonschema/1-0-0", new Object()).toString());
    }

}

vagrant/push.bash
#!/bin/bash
set -e

# Constants
bintray_user=deploy-snowplow-generic


# Similar to Perl die
function die() {
    echo "$@" 1>&2 ; exit 1;
}

# Check if our Vagrant box is running. Expects `vagrant status` to look like:
#
# > Current machine states:
# >
# > default                   poweroff (virtualbox)
# >
# > The VM is powered off. To restart the VM, simply run `vagrant up`
#
# Parameters:
# 1. out_running (out parameter)
function is_running {
    [ "$#" -eq 1 ] || die "1 argument required, $# provided"
    local __out_running=$1

    set +e
    vagrant status | sed -n 3p | grep -q "^default\s*running (virtualbox)$"
    local retval=${?}
    set -e
    if [ ${retval} -eq "0" ] ; then
        eval ${__out_running}=1
    else
        eval ${__out_running}=0
    fi
}

# Get version, checking we are on the latest
#
# Parameters:
# 1. out_version (out parameter)
# 2. out_error (out parameter)
function get_version {
    [ "$#" -eq 2 ] || die "2 arguments required, $# provided"
    local __out_version=$1
    local __out_error=$2

    # get the version number in a line like 'version = "0.1.0"' from Cargo.toml (e.g 0.1.0)
    file_version="$(vagrant ssh -c 'cd /vagrant && gradle printVersion --quiet' | perl -ne 'print $1 if /(\d+\.\d+\.\d+)/')"
    # take the latest tag, stripping anything not looking like a version (e.g. 0.1.0-rc99 => 0.1.0)
    tag_version="$(git describe --abbrev=0 --tags | perl -ne 'print $1 if /(\d+\.\d+\.\d+)/')"
    
    tag_name="$(git describe --abbrev=0 --tags)"
    
    if [[ "${file_version}" != "${tag_version}" ]]; then
        eval ${__out_error}="'File version ${file_version} != tag version ${tag_version}'"
    else
        eval ${__out_version}=${tag_name}
    fi
}

# Go to parent-parent dir of this script
function cd_root() {
    source="${BASH_SOURCE[0]}"
    while [ -h "${source}" ] ; do source="$(readlink "${source}")"; done
    dir="$( cd -P "$( dirname "${source}" )/.." && pwd )"
    cd ${dir}
}



cd_root

# Precondition for running
running=0 && is_running "running"
[ ${running} -eq 1 ] || die "Vagrant guest must be running to push"

# Precondition
version="" && error="" && get_version "version" "error"
[ "${error}" ] && die "Versions don't match: ${error}. Are you trying to publish an old version, or maybe on the wrong branch?"

# get the api key
read -es -p "Please enter API key for Bintray user ${bintray_user}: " bintray_api_key
echo

vagrant ssh -c "cd /vagrant && ./gradlew bintrayUpload -PBINTRAY_USER=${bintray_user} -PBINTRAY_KEY=${bintray_api_key}"



deploy/deploy.py
import yaml
import json
import os
import subprocess
import sys

class Executor(object):

    def exec_output(self, args):
        """Execute the arguments in args, return stdout as a string"""
        return subprocess.check_output(args)
        
    def call(self, args):
        """Execute the arguments in args, return an integer return code"""
        FNULL = open(os.devnull, 'w')
        return subprocess.call(args, stdout=FNULL, stderr=subprocess.STDOUT)
        
class Configuration(object):

    def __init__(self, config_file):
        """Parse the configuration file and put it in conf"""
        self.conf = Configuration.parse_file(config_file)

    @staticmethod
    def parse_file(config_file):
        """Read the configuration file as a string, call parse_str to return a configuration dictionary"""
        with open(config_file) as f: conf = Configuration.parse_str(f)
        return conf
        
    @staticmethod
    def parse_str(str):
        """Read the configuration file as a YAML string, return a configuration dictionary. Validate along the way"""
        c = yaml.load(str)
        if c == None:
            raise ValueError("Configuration is empty")
        
        if not 'snowplow' in c:
            raise ValueError("Configuration is missing 'snowplow' fields")
        if c['snowplow'] == None or not 'collector' in c['snowplow']:
            raise ValueError("Configuration is missing 'collector' field under 'snowplow'")
        if c['snowplow']['collector'] == None:
            raise ValueError("Configuration does not have a valid collector url set")
        if 'app_id' not in c['snowplow'] or c['snowplow']['app_id'] == None:
            raise ValueError("Configuration does not have a valid app id set")
        
        if not 's3' in c:
            raise ValueError("Configuration is missing 's3' fields")
        if c['s3'] == None or not 'buckets' in c['s3']:
            raise ValueError("Configuration is missing 'buckets' field under 's3'")
        
        if c['s3']['buckets'] == None:
            raise ValueError("No buckets specified in config!") 
        
        buckets = c['s3']['buckets']
                
        c['s3']['buckets'] = [x for x in buckets if x is not None]   
        
        if len(c['s3']['buckets']) == 0:
            raise ValueError("No buckets specified in config!")
            
        return c
                      
class Shell(object):  

    @staticmethod  
    def contains_aws_cli(executor=Executor()): 
        """Check that AWS-CLI is on the path using which, return True if it is (false if not)"""
        try: 
            executor.call(['which', 'aws-cli'])
            return True
        except subprocess.CalledProcessError:
            return False

class S3(object):
    
    def __init__(self, executor=Executor()):
        """Load the available buckets using into available_buckets using AWS-CLI"""
        self.available_buckets = S3.get_buckets_from_list(S3.get_bucket_list(executor))
    
    @staticmethod
    def get_buckets_from_list(json_str):
        """Parses a json list of available buckets without any prefix (no s3://) into a list of buckets"""
        buckets=json.loads(json_str)
        return buckets         

    @staticmethod
    def get_bucket_list(executor=Executor()):
        """Uses aws s3api to get a list of available buckets as JSON"""
        avail_buckets = executor.exec_output(['aws', 's3api', 'list-buckets', '--query', 'Buckets[].Name']) 
        return avail_buckets
        
    @staticmethod
    def add_bucket_notifications(lambda_function_name, lambda_arn, s3_bucket_names, source_account_number, executor=Executor()):
        """Add a notification to each bucket in s3_bucket_names such that the lambda_arn will be invoked for S3 get and S3 put events
        Uses AWS s3api to perform this action
        """
        if len(s3_bucket_names) == 0:
            raise ValueError("cannot add bucket notifications: 0 buckets requested")

        notification_config="{\"CloudFunctionConfiguration\":{\"CloudFunction\":\"" + lambda_arn + "\",\"Events\":[\"s3:ObjectCreated:*\",\"s3:ObjectRemoved:*\"]}}"
        
        for idx,bucket in enumerate(s3_bucket_names):
            executor.call(['aws', 'lambda', 'add-permission', '--function-name', lambda_function_name, '--action', 'lambda:InvokeFunction', 
                            '--principal', 's3.amazonaws.com', '--source-arn', 'arn:aws:s3:::' + bucket, '--source-account', source_account_number, '--statement-id', 'Id-' + str(idx+1) ]) 
            executor.call(['aws', 's3api', 'put-bucket-notification', '--bucket', bucket, '--notification', notification_config])
    
class AWS(object): 

    @staticmethod
    def get_account_id(executor=Executor()):
        """Get the users current AWS account ID 
        This is collected using the AWS-CLI - by describing the default ec2 security group
        """
        account_id = executor.exec_output(['aws','ec2', 'describe-security-groups', '--group-names', 'Default', '--query', 'SecurityGroups[0].OwnerId', '--output', 'text'])
        return account_id.rstrip()
        
class IAM(object):

    @staticmethod
    def create_lambda_policy(filename, executor=Executor()):
        """Creates an IAM policy using the specified policy file in filename"""
        policy_res = executor.exec_output(['aws', 'iam', 'create-policy', '--policy-name', 'snowplow-s3-source', '--policy-document', 'file://' + filename])
        return policy_res
    
    @staticmethod
    def create_lambda_role(role_name, filename_trust_policy, executor=Executor()):
        """Creates an IAM role as role_name using the filename_trust_policy role policy document"""
        role_result = executor.exec_output(['aws', 'iam', 'create-role', '--role-name', role_name, '--assume-role-policy-document', 'file://' + filename_trust_policy])
        return role_result
    
    @staticmethod
    def get_arn_from_policy(policy):
        """Given the policy JSON returned by AWS-CLI, extract the ARN given to the policy and return it as a string"""
        pol = json.loads(policy)
        arn = pol['Policy']['Arn']
        return arn
        
    @staticmethod
    def get_arn_from_role(role):
        """Given the role JSON returned by AWS-CLI, extract the ARN given to the role and return it as a string"""
        r = json.loads(role)
        arn = r['Role']['Arn']
        return arn    
    
    @staticmethod
    def attach_policy_to_role(role_name, policy_arn, executor=Executor()):
        """Connect an IAM policy to a role using the AWS-CLI"""
        executor.call(['aws', 'iam', 'attach-role-policy','--role-name='+role_name, '--policy-arn', policy_arn])

class Lambda(object):

    @staticmethod
    def create_lambda_function(lambda_name, lambda_zip_filename, role_arn, collector_url, executor=Executor()):
        """Create a lambda function, lambda_name and upload the zip file containing the function.
        Also give the newly created lambda the IAM role given in role_arn, and set the description
        to collector_url
        """
        return executor.exec_output(['aws', 'lambda', 'create-function', '--function-name', lambda_name, '--runtime',
                                     'java8', '--role', role_arn, '--handler', 'com.snowplowanalytics.sources.awslambda.Main::handleRequest',
                                     '--zip-file', 'fileb://' + lambda_zip_filename, '--description', ' ' + collector_url, '--timeout', '120', '--memory-size', '512'])
        
    @staticmethod    
    def get_lambda_function_arn(lambda_json):
        """Given the AWS Lambda result json returned by AWS-CLI, extract the lambda's ARN and return it as a string"""
        r = json.loads(lambda_json)
        function_arn = r['FunctionArn']
        return function_arn

def get_lambda_zip_filename(args=[]):
    """Get the best candidate for a lambda zip path
    If there's a path in args[1], use that in preference to the current directory
    If there's nothing in args[1], look for a zip in the current directory (this will be the case in a download bundle)
    Throw an error if a good candidate cannot be found
    """
    if len(args) == 2:
        if os.path.isfile(args[1]):
            if args[1].endswith('.zip'):
                lambda_zip_filename = args[1]
            else: 
                raise ValueError("Lambda zip argument does not end with zip!")
        else:
            raise ValueError("Lambda zip argument does not exist or could not be read!")
    else:   
        zip_files = []

        for file in os.listdir("."):
            if file.endswith(".zip"):
                zip_files.append(file)
                
        if len(zip_files) > 1:
            raise ValueError("Couldn't establish which file is the lambda zip - many zips exist")
        elif len(zip_files) == 0:
            raise ValueError("Couldn't find lambda zip file in current directory")
        else:
            lambda_zip_filename = zip_files[0]
            
    return lambda_zip_filename
 
def make_lambda_config_desc(collector_url, app_id):
    """Make a configuration json for inserting into the lambda's description
    In the form {"collector_url":"abc", "app_id":"an_app_id"}"""
    
    if collector_url == None or app_id == None:
        raise ValueError("collector_url and app_id are required fields!")
    
    return "{\"collector_url\":\"" + collector_url + "\",\"app_id\":\"" + app_id + "\"}"
 
if __name__ == '__main__':
        
    if Shell.contains_aws_cli() == False:
        print "This script requires the AWS CLI to be installed and configured."
        print "See http://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-set-up.html for more information"
        exit(1)
    
    lambda_zip_filename = ''
    try:
        lambda_zip_filename = get_lambda_zip_filename(sys.argv)
    except ValueError as e:
        print "Couldn't get lambda zip file: " + str(e)
        exit(1)  
        
    print "Using lambda function zip", lambda_zip_filename
    
    try:
      config = Configuration('config.yaml').conf
    except IOError as (_, strerror):
        print "Couldn't read file 'config.yaml': " + strerror 
        exit(1)
    except ValueError as e:
        print "Configuration file 'config.yaml' is invalid: " + str(e)
        exit(1)
    except Exception as e:
        print "Couldn't read configuration file 'config.yaml': " + str(e)
        exit(1)  

    existing_buckets = S3().available_buckets
    if existing_buckets == None or len(existing_buckets) == 0:
        raise ValueError("No existing buckets?")
        
    for bucket in config['s3']['buckets']:
        if not bucket in existing_buckets:
            print "Bucket '" + bucket + "' in config.yaml does not appear to exist!"
            print "\nAvailable buckets:\n"
            print "\n".join(existing_buckets)
            exit(1)
    
    print "Getting AWS account ID..."
    aws_account_id = AWS.get_account_id()
    print "Success: AWS account ID " + aws_account_id
    
    print "Setting up IAM policy for lambda function..."
    lambda_policy_arn = IAM.get_arn_from_policy(IAM.create_lambda_policy('perms.json'))
    print "Success: created lambda IAM policy as " + lambda_policy_arn
    
    print "Setting up IAM role for lambda function..."
    lambda_role_name = 'snowplow-s3-source'
    lambda_role_arn = IAM.get_arn_from_role(IAM.create_lambda_role(lambda_role_name,'lambda_trust_policy.json'))
    print "Success: created role " + lambda_role_name + " with arn " + lambda_role_arn
    
    print "Linking policy to role..."
    IAM.attach_policy_to_role(lambda_role_name, lambda_policy_arn)
    print "Success"
    
    print "Uploading lambda function..."
    lambda_name = 'snowplow-aws-lambda'

    lambda_desc = make_lambda_config_desc(config['snowplow']['collector'], config['snowplow']['app_id'])
    lambda_function_arn = Lambda.get_lambda_function_arn(Lambda.create_lambda_function(lambda_name, lambda_zip_filename, lambda_role_arn, lambda_desc))
    print "Success: uploaded lambda " + lambda_name + " with arn " + lambda_function_arn
     
    print "Adding notification rules to buckets..."    
    S3.add_bucket_notifications(lambda_name, lambda_function_arn, config['s3']['buckets'], aws_account_id)
    print "Success: buckets " + ",".join(config['s3']['buckets']) + " configured to notify AWS lambda on created/deleted events"
    
    print "Complete!"
    